### PR TITLE
improved support for Google Cloud SQL

### LINKED ISSUE #11414: Support Postgres connection via UNIX socket that contain ":" for application DB (enables easy Google Cloud Run deployment)
**Is your feature request related to a problem? Please describe.**
I am trying to run Metabase on Google Cloud Run, using Postgres (via Cloud SQL) as the application database instead of H2. The problem is that Cloud Run only supports connections through Unix sockets. While I believe Unix sockets work just fine otherwise, in the particular case of Cloud SQL the socket name contains a ":", which breaks the regex used on MB_DB_CONNECTION_URI.

Cloud Run just entered GA and looks to be the cheapest option available for hosting Metabase on GCP. Running Metabase on App Engine still requires some hacks, so Cloud Run has the potential to be the cheapest AND easiest hosting option on GCP... with this one fix!

**Describe the solution you'd like**
Let us specify a %-encoded MB_DB_CONNECTION_URI to connect to Postgres via Unix socket that contains ":" characters.

**How important is this feature to you?**
I'm tied to GCP. App Engine Flex seems like overkill for a service used by 2-3 internal users max. Anything else requires more devops work than i care to undertake. So, a minor change would make it really easy (and cheap!) to work with Metabase.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by jkeys089**
Support for directly connecting to Google Cloud SQL db from serverless environments.

### CODE REVIEW

**Comment by jkeys089** [comment]
@flamber Happy New Year! Just checking in to see if this is something you might be interested in merging? If so, I'm happy to make any improvements that are needed.

**Comment by flamber** [comment]
Hi @jkeys089 
Not sure what else should be done, but perhaps also add support for MySQL? #3852

**Review by camsaul — COMMENTED** [review]
Let me take a crack at implementing https://github.com/metabase/metabase/issues/19961 real quick which would eliminate the need for you to change the regex or actually do anything other than add that dep.

What's the license for the new dep?

**Comment by jkeys089** [comment]
> Hi @jkeys089 Not sure what else should be done, but perhaps also add support for MySQL? #3852

Sure -- the PR has been updated to include MySQL as well.

**Comment by jkeys089** [comment]
> Apache 2.0 according to https://mvnrepository.com/artifact/com.google.cloud.sql/postgres-socket-factory

That's correct (for both postgres and mysql): https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/LICENSE

**Review by camsaul — CHANGES_REQUESTED** [review]
Now that #19970 can you update this to fix the merge conflicts? Since we're not parsing the connection string at all anymore you should be able to remove all those changes entirely.

File: deps.edn
Comment (camsaul) [inline deps.edn:40]: Looks like `1.4.3` is out now

**Review by camsaul — CHANGES_REQUESTED** [review]
Have you tested this locally and confirmed it actually works? The more I look at it the more I'm convinced it won't actually work, for a couple of reasons.

1. This seems like this only works with the MySQL JDBC driver:

   > Socket factory for the MySQL JDBC driver (version 8.x)

   We don't use the MySQL JDBC driver, we use the MariaDB JDBC driver, and we can't use the MySQL JDBC driver because it's GPL licensed and would be incompatible with our Enterprise license. Either way, would it still work with the MariaDB driver? If you look at [the actual SocketFactory impl added by the new dep](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/jdbc/mysql-j-8/src/main/java/com/google/cloud/sql/mysql/SocketFactory.java#L33) it looks like it extends `com.mysql.cj.protocol.SocketFactory` which [definitely doesn't exist in the MariaDB JDBC driver](https://github.com/mariadb-corporation/mariadb-connector-j/tree/b4fa9705d242b0c566a87f8c6c82fb5a3982d627/src/main/java).
 
2. [If you don't specify a `host` we default to `localhost`](https://github.com/metabase/metabase/blob/master/src/metabase/db/spec.clj#L43-L51). Actually with 0.43.0 and #19961 we wouldn't merge in the default host if you use `MB_DB_CONNECTION_URI` for the application DB connection. But it still wouldn't work for anything else.)

Either way since [you have to include additional JDBC options to get this to work](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/docs/jdbc-mysql.md) we should probably mention this in the dox as well. Otherwise people are going to open bug reports about how we merged this issue to add Google Cloud Platform support but it doesn't work out of the box

**Comment by jkeys089** [comment]
> Have you tested this locally and confirmed it actually works? The more I look at it the more I'm convinced it won't actually work, for a couple of reasons.

No, I have not tested the mysql driver in a live env. We're using this with postgres in a live Google App Engine Flexible environment. I added mysql at the request of flamber. I'd be happy to setup a quick test project to verify once we settle on a direction for mysql.

> 1. This seems like this only works with the MySQL JDBC driver:
>    > Socket factory for the MySQL JDBC driver (version 8.x)
>    
>    
>    We don't use the MySQL JDBC driver, we use the MariaDB JDBC driver, and we can't use the MySQL JDBC driver because it's GPL licensed and would be incompatible with our Enterprise license. Either way, would it still work with the MariaDB driver? If you look at [the actual SocketFactory impl added by the new dep](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/jdbc/mysql-j-8/src/main/java/com/google/cloud/sql/mysql/SocketFactory.java#L33) it looks like it extends `com.mysql.cj.protocol.SocketFactory` which [definitely doesn't exist in the MariaDB JDBC driver](https://github.com/mariadb-corporation/mariadb-connector-j/tree/b4fa9705d242b0c566a87f8c6c82fb5a3982d627/src/main/java).

Understood. Perhaps it would be best to just remove mysql then? It seems to be incompatible at a fairly fundamental level. How do you feel about just supporting postgres and not mysql for GCP? I'm guessing that might be a deal breaker if you want to maintain feature parity for all supported application db's.

> 2. [If you don't specify a `host` we default to `localhost`](https://github.com/metabase/metabase/blob/master/src/metabase/db/spec.clj#L43-L51). Actually with 0.43.0 and #19961 we wouldn't merge in the default host if you use `MB_DB_CONNECTION_URI` for the application DB connection. But it still wouldn't work for anything else.)

That's actually OK. The [host is ignored](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/docs/jdbc-postgres.md#creating-the-jdbc-url):

> Note: The host portion of the JDBC URL is currently unused, and has no effect on the connection process. The SocketFactory will get your instances IP address based on the provided cloudSqlInstance arg.

The [sample jdbc url's](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/docs/jdbc-postgres.md#creating-the-jdbc-url) provided by Google don't include a host so this change was meant to accommodate that.

> Either way since [you have to include additional JDBC options to get this to work](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/docs/jdbc-mysql.md) we should probably mention this in the dox as well. Otherwise people are going to open bug reports about how we merged this issue to add Google Cloud Platform support but it doesn't work out of the box

Sure. I'd be happy to add to add some documentation / examples.

**Comment by camsaul** [comment]
@jkeys089 @flamber since MySQL isn't going to work for the reasons mentioned above I think just doing this for Postgres is fine for now. It would be nice to do it for MySQL as well but I don't see a way to make it work.

@flamber there's also a SQL Server version of this but I'll leave it up to you whether we want to put that in here

**Comment by flamber** [comment]
Yeah, let's land this for Postgres only - seems like most uses Postgres.

File: deps.edn
Comment (jkeys089) [reply deps.edn]: Updated.

**Review by camsaul — APPROVED** [review]

**Comment by camsaul** [comment]
Cool, this LGTM. I just triggered a CI run, a few unrelated tests should fail since we're not injecting env var info but if everything else passes we can merge this in

**Comment by yi-jiayu** [comment]
Hi, apologies for bumping this PR. Are there any plans to merge this in the near future? I think having `com.google.cloud.sql/postgres-socket-factory` as a built-in dependency would be really useful.

Thanks for building Metabase!

**Comment by yi-jiayu** [comment]
Hi, apologies for bumping this PR. Are there any plans to merge this in the near future? I think having `com.google.cloud.sql/postgres-socket-factory` as a built-in dependency would be really useful.

Thanks for building Metabase!

**Comment by Kulikovpavel** [comment]
Are there any plans to merge this in the near future?

**Comment by Kulikovpavel** [comment]
@camsaul ?

**Comment by camsaul** [comment]
Ok I fixed the merge conflicts and pushed PR #22923 up so it runs against CI -- if that's green then we can merge

**Comment by camsaul** [comment]
Looks like it's failing CI because it can't detect the licenses for the new dependencies. They'll have to be added to https://github.com/metabase/metabase/blob/master/bin/build-mb/resources/overrides.edn

**Comment by jkeys089** [comment]
> Looks like it's failing CI because it can't detect the licenses for the new dependencies. They'll have to be added to https://github.com/metabase/metabase/blob/master/bin/build-mb/resources/overrides.edn

@camsaul is this something you want me to add? If so, should I add it in this PR or https://github.com/metabase/metabase/pull/22923?

**Comment by camsaul** [comment]
@jkeys089 you can add it here and I'll pull it in to #22923 to verify that CI is passing. Then we can merge this one if you want to fix the merge conflicts here as well

You can test it locally with 

```sh
cd bin/build-mb
clojure -M:test
```

and that will pass once the license info is added manually

**Comment by jkeys089** [comment]
@camsaul it is still reporting a missing lic for `com.github.jnr/jffi` because the jar contains no pom.xml. The license info is inside the jar in `META-INF/MANIFEST.MF`:
```
Manifest-Version: 1.0
Archiver-Version: Plexus Archiver
Created-By: Apache Maven Bundle Plugin
Automatic-Module-Name: org.jnrproject.jffi.nativelibs
Bnd-LastModified: 1641486857425
Build-Jdk: 11.0.4
Built-By: headius
Bundle-ClassPath: .,jni
Bundle-Description: Java Foreign Function Interface - Native Libraries
Bundle-License: http://www.apache.org/licenses/LICENSE-2.0.txt
Bundle-ManifestVersion: 2
Bundle-Name: jffi
Bundle-SymbolicName: com.github.jnr.jffi.native
Bundle-Version: 1.3.9
Fragment-Host: com.github.jnr.jffi
Require-Bundle: com.github.jnr.jffi
Tool: Bnd-2.1.0.20130426-122213
```

I guess we could hard-code the pom.xml from https://search.maven.org/artifact/com.github.jnr/jffi/1.3.9/jar as a string? I'm not sure if there is a better way of handling this.

**Comment by jkeys089** [comment]
@camsaul I took a shot at fixing the license check. I guess the entire thing should eventually be updated to use clojure deps but that should probably be handled by someone inside MB 😄

**Comment by Kulikovpavel** [comment]
Any news here, @camsaul ?

**Comment by dpsutton** [comment]
The license check seems fine. As soon as this lands I'll clean it all up and use the basis rather than the classpath now that we are based on deps.edn and not lein.

**Comment by dpsutton** [comment]
@jkeys089 I cleaned up our licensing code so it won't break if there is no pom. If you revert your changes to build.clj and licenses.clj and rebase on master it should work for you.

**Comment by jkeys089** [comment]
@dpsutton thanks! Everything seems to be passing now.

**Comment by camsaul** [comment]
I pushed the latest version of this PR to #22923 so we can test it against CI. If everything passes there then we can go ahead and merge this.

**Review by camsaul — APPROVED** [review]

**Comment by alexphilpro** [comment]
@camsaul any updates on pushing that forward?

**Comment by camsaul** [comment]
Looks like CI is passing in #22923 except for an unrelated dependency scan error so we can go ahead and merge this as soon as the merge conflicts get resolved

**Comment by camsaul** [comment]
I went ahead and fixed the merge conflicts. Will merge this now

**Comment by flamber** [comment]
@jkeys089 Just FYI, we had to remove this, since it broke BigQuery #23976

**Comment by jkeys089** [comment]
@flamber I suppose this was the [same issue](https://discourse.metabase.com/t/bigquery-noclassdeffounderror/20250) we ran into. The resolution for us was to include the bigquery driver directly in deps.edn just before the postgres socket factory.

**Comment by flamber** [comment]
@jkeys089 We cannot do that. The drivers has to be modular, so we'll have to figure out another solution to support this.

**Pull request merged**
